cloud/aws/services/ec2/natgateways.go

```diff
@@ -0,0 +1,110 @@
+// Copyright © 2018 The Kubernetes Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ec2
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api-provider-aws/cloud/aws/providerconfig/v1alpha1"
+)
+
+func (s *Service) reconcileNatGateways(subnets []*v1alpha1.Subnet, vpc *v1alpha1.VPC) error {
+	if !s.containsPrivateSubnets(subnets) {
+		return nil
+	}
+
+	existing, err := s.describeNatGatewaysBySubnet(vpc.ID)
+	if err != nil {
+		return err
+	}
+
+	for _, sn := range subnets {
+		if sn.ID == "" || !sn.IsPublic {
+			continue
+		}
+		if _, ok := existing[sn.ID]; ok {
+			continue
+		}
+
+		_, err := s.createNatGateway(sn.ID)
```

Comment:
Do we want to create a new nat gateway for each public subnet? Or one per AZ?

Reply:
For MVP, I think we should pick sensible defaults for subnet size, and then if you want to add a subnet to an existing AZ within the same cluster, it's because of scalability - and then you probably want an additional NAT gateway on top anyway.

I imagine there to be more variations down the lines. Within our customer base, they vary from:

* Using EC2 instead of NAT Gateway for to reduce data transfer cost purposes.
* Transparent proxying, with a mix of block & allow lists, TLS intercept, ClamAV scanning etc... (try to push back on this stuff as much as possible, but don't always win out)
* DNS intercept

These EC2-based gateways can be handled with passing in an AMI ID and starting a machine appropriately, but this relates to #88 .

Reply:
I share the sentiment of picking sensible defaults. What do we think about creating only a single nat gateway for now and share it across the subnets in the same AZ? This would be good to track in a separate issue for long term plans. 

MVP is single AZ, see https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/docs/proposal/mvp-architecture.md.

Reply:
Tracked here: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/89